src/Interpreters/TreeRewriter.cpp

```diff
@@ -244,6 +244,32 @@ void renameDuplicatedColumns(const ASTSelectQuery * select_query)
     }
 }
 
+/// Aliases in WITH clause might not be referred by anyone. Remove them to avoid redundant analysis.
+void removeUnusedAliasesFromWithClause(ASTSelectQuery * select_query)
+{
+    auto new_with = select_query->with();
+
+    if (select_query->with())
+    {
+        ASTs & elements = select_query->with()->children;
+        elements.erase(
+            std::remove_if(
+                elements.begin(),
+                elements.end(),
+                [](const ASTPtr & elem)
+                {
+                    if (auto * ast_with_alias = dynamic_cast<ASTWithAlias *>(elem.get()))
+                        return !ast_with_alias->used;
+                    else
+                        return true; // ASTWithElement
+                }),
+            elements.end());
```

Comment:
In c++20 it can be rewritten using `std::erase_if`, i.e.:

```suggestion
        std::erase_if(elements, [](const ASTPtr & elem)
        {
            if (auto * ast_with_alias = dynamic_cast<ASTWithAlias *>(elem.get()))
                return !ast_with_alias->used;
            else
                return true; // ASTWithElement
        });
```

Reply:
FWIW looks like arcadia is fine with this c++20 change